### ISSUE TITLE
"Cannot find install-sh, install.sh, or shtool"

### ISSUE THREAD

**Issue by fredrik-johansson** [flintlib/flint#1386]
Had this error running ``./configure`` on a 32-bit Linux vm image from oxboxes.org. The solution in https://askubuntu.com/questions/27677/cannot-find-install-sh-install-sh-or-shtool-in-ac-aux of running

    libtoolize --force
    aclocal
    autoheader
    automake --force-missing --add-missing
    autoconf

seems to do the trick. Probably something in the config script should be tweaked so that no workarounds are needed.

**Comment by fredrik-johansson** [comment]
Comments from Andreas Enge:

> what is missing is a call to "automake". I did this:
$ automake --add-missing
[configure.ac](http://configure.ac/): error: no proper invocation of AM_INIT_AUTOMAKE was found.
[configure.ac](http://configure.ac/): You should verify that [configure.ac](http://configure.ac/) invokes AM_INIT_AUTOMAKE,
[configure.ac](http://configure.ac/): that aclocal.m4 is present in the top-level directory,
[configure.ac](http://configure.ac/): and that aclocal.m4 was recently regenerated (using aclocal)
[configure.ac:100](http://configure.ac:100/): installing './config.guess'
[configure.ac:100](http://configure.ac:100/): installing './config.sub'
[configure.ac:465](http://configure.ac:465/): installing './install-sh'
automake: error: no 'Makefile.am' found for any configure output

> Normally autoreconf would do it, but since you are effectively not using
automake, I think it does not.

> In the short run, I would add "automake --add-missing" to the bootstrap
script. And to keep things a bit tidy, I would add
AC_CONFIG_AUX_DIR([config])
to [configure.ac](http://configure.ac/). This will cause all the messy extra files to be put into
a config subdirectory.

> In the long run, you may switch to using automake. This probably means
more black magic, but maybe also a simpler [configure.ac](http://configure.ac/).

@albinahlback What do you think?

**Comment by albinahlback** [comment]
Does adding `automake --add-missing` and `aclocal` solve the problem? What versions of autotools are being shipped with that version?

**Comment by AndreasEnge** [comment]
The first attached patch solves the problem.

The second one is unrelated, but cleans up the main directory a bit by putting the helper files into the config/ subdirectory.

[0001-Add-a-missing-call-to-automake-to-bootstrap.sh.txt](https://github.com/flintlib/flint2/files/11496596/0001-Add-a-missing-call-to-automake-to-bootstrap.sh.txt)
[0002-Add-a-configuration-directory-to-configure.ac.txt](https://github.com/flintlib/flint2/files/11496597/0002-Add-a-configuration-directory-to-configure.ac.txt)

**Comment by fredrik-johansson** [comment]
@AndreasEnge with your patches, I get

```
fredrik@mordor:~/src/flint2$ ./bootstrap.sh 
autoreconf: export WARNINGS=all
autoreconf: Entering directory '.'
autoreconf: configure.ac: not using Gettext
autoreconf: running: aclocal 
autoreconf: configure.ac: tracing
autoreconf: configure.ac: creating directory config
autoreconf: running: libtoolize
libtoolize: putting auxiliary files in AC_CONFIG_AUX_DIR, 'config'.
libtoolize: linking file 'config/ltmain.sh'
libtoolize: Consider adding 'AC_CONFIG_MACRO_DIRS([m4])' to configure.ac,
libtoolize: and rerunning libtoolize and aclocal.
libtoolize: Consider adding '-I m4' to ACLOCAL_AMFLAGS in Makefile.am.
autoreconf: configure.ac: not using Intltool
autoreconf: configure.ac: not using Gtkdoc
autoreconf: running: aclocal 
autoreconf: running: /usr/bin/autoconf
autoreconf: running: /usr/bin/autoheader
autoreconf: configure.ac: not using Automake
autoreconf: linking config/config.sub to /usr/share/autoconf/build-aux/config.sub
autoreconf: linking config/config.guess to /usr/share/autoconf/build-aux/config.guess
autoreconf: linking config/install-sh to /usr/share/autoconf/build-aux/install-sh
autoreconf: Leaving directory '.'
configure.ac: error: no proper invocation of AM_INIT_AUTOMAKE was found.
configure.ac: You should verify that configure.ac invokes AM_INIT_AUTOMAKE,
configure.ac: that aclocal.m4 is present in the top-level directory,
configure.ac: and that aclocal.m4 was recently regenerated (using aclocal)
automake: error: no 'Makefile.am' found for any configure output

```

**Comment by AndreasEnge** [comment]
@fredrik-johansson Yes, there is an error message, but the call nevertheless adds install-sh.

@albinahlback I do not know why we need to call automake to install that file that apparently is needed by the configure script. But it works... Alternatively, you could check a copy into the git repository.

**Comment by fredrik-johansson** [comment]
Can the error message be suppressed though?

We'll get a lot of confused users when the last output of ``./bootstrap.sh`` is an error.

**Comment by AndreasEnge** [comment]
In the file, one could do
   automake --add-missing 2> /dev/null
which is dangerous because it might hide real problems.
Apart from that I have not found a solution.

**Comment by AndreasEnge** [comment]
I have tested with autoconf 2.69 and 2.71 and automake 1.16.5, all from the GNU Guix distribution.

**Comment by edgarcosta** [comment]
I cannot add much, as I don't understand these kinds of tools, but I observed the same when trying to build flint with GitHub actions on ubuntu-20.04:
https://github.com/assaferan/omf5/actions/runs/5164077658/jobs/9302765372#step:5:25

**Comment by fredrik-johansson** [comment]
Possibly fixed now.

**Comment by albinahlback** [comment]
I now have this issue on Debian Buster.

**Comment by fingolfin** [comment]
Failure to copy config.guess and config.sub (and `install-sh` etc.) is a known bug in autoconf 2.69, which is fixed in 2.71 (and perhaps 2.70, but I never tried with that).

I've used workarounds similar to what @AndreasEnge suggested above before, see e.g. here: https://github.com/semigroups/Semigroups/blob/main/autogen.sh

**Comment by albinahlback** [comment]
Should be fixed by #1446
